MCDataProducts/inc/DigiProvenance.hh

```diff
@@ -0,0 +1,32 @@
+// Flag to designate a digi as having been produced from simulation,
+// read from preexisting data, or a hybridization fo the two
+// Ed Callaghan, 2024
+
+#ifndef MCDataProducts_DigiProvenance_hh
+#define MCDataProducts_DigiProvenance_hh
+
+// stl
+#include <map>
+#include <string>
+
+// mu2e
+#include "Offline/GeneralUtilities/inc/EnumToStringSparse.hh"
+
+namespace mu2e{
+  // enum equipped with std::string descriptions
+  class DigiProvenanceDetail{
+    public:
+      enum enum_type {unknown=0, Simulation, Mixed, External};
+      static std::string const& typeName();
+      static std::map<enum_type, std::string> const& names();
+  };
+  using StringedDigiProvenance = EnumToStringSparse<DigiProvenanceDetail>;
+
+  class DigiProvenance: public StringedDigiProvenance{
```

Comment:
Three things. First, EnumToStringSparse does not have a virtual d'tor so we should not inherit from it. That's easy to fix, we can give it a virtual d'tor and also add the other 4 rule of 5 functions as =default.  ( My style is if the compiler will do the right thing for all rule of 5 functions I just leave them out and add a one line comment to that effect  - I see that EnumToStringSparse is so old that the rule of 5 was then the rule of 3.)  Go ahead and edit EnumToStringSparse.

The second issue is the CSAID C++ experts strongly advise against using inheritance when containment will do.  We do violate this in some places.  I think we are OK here.

Third,  I think that if you add
typedef StringedDigiProvenance::enum_type enum_type;

(or try the equivalent using would be better ).

Then downstream code can avoid needing the DigiProvenanceDetail:: and you may also be able to avoid the static casts. I am not 100% sure that I have all details right here but I am pretty sure I am close.

Reply:
After sleeping on it, there is an different solution that avoids introducing the third type and avoids issue with inheritance, just use a free function:

```
using DigiProvenance = EnumToStringSparse<DigiProvenanceDetail>;

bool containsSimulation(  DigiProvenance::enum_type  id ){
   return (  id == DigiProvenance::Simulation || id == DigiProvenance::Mixed );
}
```

Reply:
The free-function approach appears to be the path of least resistance --- it is preferable, IMO, to the extra baggage needed to make another distinct class pass muster. (FWIW, I went with inheritance over containment here because otherwise there are extra implementations to add to the list (i.e. forwarding all of the member functions of `EnumToStringSparse<>`).

That being said, I'm seeing issues with serialization of that simpler implementation (`using DigiProvenance = EnumToStringSparse<DigiProvenanceDetail>`) manifesting as data corruption after deserializing, in which the actual `enum_type` not being set correctly. Since this should be simpler the scenario, I'm not really sure what's going wrong. I'll have another try at figuring it out, but if I can't then I may need to just push updates to `EnumToStringSparse<>` to make it "inheritable."

Reply:
Sorry for the slow response.  

Can you point me to your working code that is failing and tell me what the error message is?

Reply:
No worries. I have the changes in a branch at https://github.com/edcallaghan/mu2e-Offline/tree/update_TrkStrawHitMC_provenance_free_functions. If you compile that (should work at head) and run the contents of `~ejc3/public/DigiProvenance`,  you can reproduce for yourself. (This is a pair of jobs exemplified in `.../run.sh`, in which we first ``physics-mix'' [pm] a few NoPrimary [np] events, and then ``digi-mix'' [dm] a few CEs into them [ce]). You can see the error in `.../dm-ce.log`, the relevant bit pasted here:
```
---- EventProcessorFailure BEGIN
  EventProcessor: an exception occurred during current event processing
  ---- ScheduleExecutionFailure BEGIN
    Path: ProcessingStopped.
    ---- ProductNotFound BEGIN
      A request to resolve an Ptr to a product containing items of type: mu2e::StrawGasStep with ProductID 2189702261
      cannot be satisfied because the product cannot be found.
      The productGetter was not set -- are you trying to dereference a Ptr during mixing?
      The above exception was thrown while processing module StrawDigiMCFilter/Triggerable run: 1202 subRun: 0 event: 1
    ---- ProductNotFound END
    Exception going through path TriggerablePath
  ---- ScheduleExecutionFailure END
---- EventProcessorFailure END
```

But: my statement that the problem is in the serialization of the provenance was confused and, I think, incorrect. After another round of less-sloppy testing, the problem presents with the current version in this PR as well.

I suspect that the actual problem is with the construction of a new `StrawDigiMC` at https://github.com/Mu2e/Offline/blob/main/TrackerMC/src/StrawDigiBundleCollection.cc#L223, ultimately stemming from the same "scope constraint" on use of art data products as I mention in https://github.com/Mu2e/Offline/pull/1290. That has nothing to do with the implementation of `DigiProvenance`, so I can push the simpler implementation for that here. If I'm right, then another PR will need to be opened which addresses the invalid `Ptr`s.

Reply:
I am looking at it now.

Just a wild guess and it's from memory.  If your are making two data products, ACollection and BCollection, both in the same module, and if one contains Ptrs into the other, you cannot dereference the Ptrs until the module has returned.  So they only work in downstream modules.  The reason is that the pointee data product does not exist until after the return ( the move and the return are like a database two phase commit ).

Reply:
Please make dm-ce.fcl readable by me. 

ls -l ~ejc3/public/DigiProvenance/dm-ce.fcl 
-rw-------+ 1 ejc3 nobody 4177 Jun 30 15:46 /nashome/e/ejc3/public/DigiProvenance/dm-ce.fcl

Reply:
It looks like my wild guess is irrelevant  The changes to Filters/src/StrawDigiMCFilter_module.cc look simple enough.  So I agree it's likely that the bug is earlier in the workflow.

That's for making the change to the free function and fixing the typo.    My reason for suggesting the free function was the same as in your reply.

Reply:
Sorry, `~ejc3/public/DigiProvenance/dm-ce.fcl` is now public-readable.

But, that being said, I think the issue here is actually much dumber than what I was getting at. When "digi mixing," I am mixing in the `StrawDigiMC`s, each of which contains an `art::Ptr<SimParticle>` into a `SimParticleCollection`. But I don't actually mix the `SimParticleCollection` in, so of course the pointer can't be dereferenced.

Reply:
Are all of the fixes in now?